classes/local/factor/object_factor_base.php

```diff
@@ -35,14 +35,30 @@ abstract class object_factor_base implements object_factor {
      */
     public $name;
 
+    /**
+     * Lock counter.
+     */
+    private $lockcounter;
+
     /**
      * Class constructor
      *
      * @param string factor name
      *
      */
     public function __construct($name) {
+        global $DB, $USER;
         $this->name = $name;
+
+        // Setup the lock counter.
+        $sql = "SELECT MAX(lockcounter) FROM {tool_mfa} WHERE userid = ? AND factor = ?";
```

Comment:
should this exclude previously revoked factors?

Reply:
I think the revoked factors should also have this field incremented. We want the lockoutcounter to be in lockstep for the entire class of factor instances, so I think that should include all of them